## CAPSTONE PROJECT-1 :- PRCP-1001-RiceLeaf

## PROJECT TEAM ID :- PTID-CDS-APR-24-1899

# 1.Business Case:
Rice plant is susceptible to diseases that may affect the quantity and quality of rice production. Detection of these diseases by the farmers will require a great knowledge about the various disease and how to identify them visually. Monitoring these diseases, their occurrences and frequencies are very important for early detection of the affected plants, their timely treatment, and most importantly, for planning future strategies to prevent the diseases to minimize the losses.
CNN is an algorithm of deep learning techniques that has been successfully invoked for handling computer vision issues such as picture classification, object segmentation, and image analysis. CNN has been used to address the classification of Rice leaf diseases accurate detection and classification of rice leaf disease.

The dataset provided contains 120 jpg images of disease-infected rice leaves. The images are grouped into 3 classes based on the type of disease. There are 40 images in each class.
Classes

* **Bacterial blight**: elongated lesions near the leaf tips and
margins, and turns white to yellow and then grey due to
fungal attack.

* **Brown spot**: 
    dark brown colored and round to oval shaped lesions on rice leaves.
    
* **Leaf smut**: small black linear lesions on leaf blades, leaf
tips may turn grey and dry.


## AIM
**Task 1:**-Prepare a complete data analysis report on the given data.

**Task 2:**-Create a model which can classify the three major attacking diseases of rice plants like leaf blast, bacterial blight and brown spot.

**Task3:**- Analyze various techniques like Data Augmentation, etc and create a report on that.

# 2.Import Libraries:

In [1]:
# Installing tensorflow:
!pip install tensorflow

In [2]:
pip install opencv-python

Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np
import pandas as pd
import seaborn as sns
import random
import matplotlib.pyplot as plt
from matplotlib.image import imread


In [5]:
import tensorflow as tf
from tensorflow import keras
from keras import layers
from keras.layers import Input , Lambda
from keras.layers import Dense, Flatten, Dropout, Conv2D, MaxPool2D, MaxPooling2D, BatchNormalization
from keras.models import Model
from keras.preprocessing import image
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from glob import glob

import warnings
warnings.filterwarnings("ignore")

ImportError: cannot import name 'ImageDataGenerator' from 'keras.preprocessing.image' (C:\Users\dell\anaconda3\Lib\site-packages\keras\api\preprocessing\image\__init__.py)

In [ ]:
# give dataset path
train_path =r'C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data\train'
test_path = r'C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data\test'

In [ ]:
# useful for getting number of classes
train_folders = glob(r'C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data\*')
print(len(train_folders))

In [ ]:
# useful for getting number of classes
test_folders = glob(r'C:\DATA SCIENCE\Capstone project\Project reference\Rice Leaf\Data\*')
print(len(train_folders))

# 3. Load Dataset: 

## Visualizing the Image Data

### 1.Bacterial leaf blight 

In [ ]:
plt.figure(figsize=(12,12))
bacteria_path = r'C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data\Bacterial leaf blight-20200814T055237Z-001\Bacterial leaf blight'
for i in range (1, 22):
    plt.subplot(7 ,3, i)
    plt.tight_layout()
    rand_imp = plt.imread(bacteria_path+'/'+random.choice(sorted(os.listdir(bacteria_path))))
    plt.imshow(rand_imp)
    plt.xlabel(rand_imp.shape[1], fontsize = 10) # width of image.
    plt.ylabel(rand_imp.shape[0], fontsize = 10) # height of image.

### 2.Brown Spot

In [ ]:
plt.figure(figsize=(12,12))
Brown_Spot = r'C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data\Brown spot-20200814T055208Z-001\Brown spot'
for i in range (1, 22):
    plt.subplot(7,3,i)
    plt.tight_layout()
    rand_imp = plt.imread(Brown_Spot+'/'+ random.choice(sorted(os.listdir(Brown_Spot))))
    plt.imshow(rand_imp)
    plt.xlabel(rand_imp.shape[1], fontsize = 10) # width of image.
    plt.ylabel(rand_imp.shape[0], fontsize = 10) # height of image.

### 3.Leaf Smut

In [ ]:
plt.figure(figsize=(12,12))
Leaf_Smut = r'C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data\Leaf smut-20200814T055530Z-001\Leaf smut'
for i in range (1, 22):
    plt.subplot(7,3,i)
    plt.tight_layout()
    rand_imp = plt.imread(Leaf_Smut+'/'+ random.choice(sorted(os.listdir(Leaf_Smut))))
    plt.imshow(rand_imp)
    plt.xlabel(rand_imp.shape[1], fontsize = 10) # width of image.
    plt.ylabel(rand_imp.shape[0], fontsize = 10) # height of image.

#  Data Preprocessing:

In [ ]:
train_datagen = ImageDataGenerator(rotation_range=40,
                                  width_shift_range=0.2,
                                  height_shift_range=0.2,
                                  rescale= 1./255,
                                  shear_range=0.2,
                                  zoom_range=0.2,
                                  horizontal_flip=True,
                                  fill_mode= "nearest",
                                  validation_split= .25)
test_datagen = ImageDataGenerator(rescale=1./255,
                                 validation_split=.25)

# Example of Data Augumentation:

In [ ]:
from tensorflow.keras.utils import img_to_array,array_to_img,load_img

In [ ]:
datagen = ImageDataGenerator(rotation_range=40,
                             width_shift_range=0.2,
                             height_shift_range=0.2,
                             shear_range=0.2,
                             zoom_range=0.2,
                             horizontal_flip=True,
                             fill_mode= "nearest")
img = load_img(r"C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data\Brown spot-20200814T055208Z-001\Brown spot\DSC_0121.jpg") # This is a PIL image.
x = img_to_array(img) # This is a Numpy array with shape (3, 150, 150)
x = x.reshape((1,)+ x.shape) # This a Numpy array with shape(1, 3, 150, 150)


# The .flow() command below generates batches of randomly transformed images
# and saves the result to the `preview/` directory

i = 0
for batch in datagen.flow(x, batch_size=1,
                         save_to_dir=r"C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Image Augmentation Example", save_prefix="rice_leaf_brown_spot", save_format="jpeg"):
    i += 1
    if i > 20:
        break  # Otherwise the generator would loop indefinitely.


In [ ]:
plt.figure(figsize=(12,12))
path = r"C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Image Augmentation Example"
for i in range(1, 13):
    plt.subplot(4, 3, i)
    plt.tight_layout()
    rand_imp = imread(path +'/'+ random.choice(sorted(os.listdir(path))))
    plt.imshow(rand_imp)
    plt.xlabel(rand_imp.shape[1], fontsize = 10) # Width of image.
    plt.ylabel(rand_imp.shape[0], fontsize = 10) # Height of image.

# Generating a Training and Validation set:

In [ ]:
training_set = train_datagen.flow_from_directory(r"C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data",
                                                target_size= (256, 256),
                                                batch_size=5,
                                                class_mode='categorical',
                                                subset='training')

In [ ]:
validation_set = train_datagen.flow_from_directory(r"C:\Users\ganes\Downloads\PRCP-1001-RiceLeaf (1)\Data",
                                                  target_size=(256, 256),
                                                  batch_size=5,
                                                  class_mode= "categorical",
                                                  subset= "validation")

# Building the Convolution Neural Network Model:

## 1. Using ADAM Optimiser:

In [ ]:
# Initialising the CNN model:
model_adam = Sequential()

# First Convolution Layer:
model_adam.add(Conv2D(filters= 16, kernel_size=2, activation="relu", input_shape=[256, 256, 3]))

# First Pooling Layer:
model_adam.add(MaxPooling2D(pool_size=2))

# Second Convolution Layer:
model_adam.add(Conv2D(filters=32, kernel_size=2, activation="relu"))

# Second Pooling Layer:
model_adam.add(MaxPooling2D(pool_size=2))

# Third Convolution Layer:
model_adam.add(Conv2D(filters=32, kernel_size=2 , activation="relu"))

# Third Polling Layer:
model_adam.add(MaxPooling2D(pool_size=2))

# Fourth Convolution Layer:
model_adam.add(Conv2D(filters=64, kernel_size=2, activation="relu"))

# Fourth Pooling Layer:
model_adam.add(MaxPooling2D(pool_size=2))

# Flattening:
model_adam.add(Flatten())


# Fully Connected Layer:
model_adam.add(Dense(128, activation="relu"))
# Deactivating 50% of neuron
model_adam.add(Dropout(0.5))


# Output Layer :
model_adam.add(Dense(3, activation="softmax"))


In [ ]:
## Layers
model_adam.layers

In [ ]:
# Summary
model_adam.summary()

In [ ]:
early_stopping = tf.keras.callbacks.EarlyStopping(monitor="val_loss",
                                                 min_delta=0.0001,
                                                 patience=20,
                                                 verbose=1,
                                                 mode= "auto",
                                                 baseline=None,
                                                 restore_best_weights=False,
                                                 start_from_epoch=0)

In [ ]:
early_stopping

## Compiling the model:


In [ ]:
from keras.optimizers import Adam

In [ ]:
model_adam.compile(loss = "categorical_crossentropy", optimizer = "adam", metrics=["accuracy"])

In [ ]:
# Training the model
history_adam = model_adam.fit_generator(training_set,
                                       steps_per_epoch=len(training_set),
                                       epochs=150,
                                       validation_data=validation_set,
                                       validation_steps=len(validation_set))

Epoch 112/150
18/18 [==============================] - 4s 214ms/step - loss: 0.1599 - accuracy: 0.9444 - val_loss: 1.5357 - val_accuracy: 0.7241
Epoch 113/150
18/18 [==============================] - 4s 222ms/step - loss: 0.2863 - accuracy: 0.8556 - val_loss: 3.3706 - val_accuracy: 0.7586
Epoch 114/150
18/18 [==============================] - 4s 212ms/step - loss: 0.4172 - accuracy: 0.8556 - val_loss: 3.4105 - val_accuracy: 0.7931
Epoch 115/150
18/18 [==============================] - 4s 210ms/step - loss: 0.3644 - accuracy: 0.8778 - val_loss: 5.9220 - val_accuracy: 0.6552
Epoch 116/150
18/18 [==============================] - 4s 218ms/step - loss: 0.3094 - accuracy: 0.9000 - val_loss: 3.5130 - val_accuracy: 0.7241
Epoch 117/150
18/18 [==============================] - 4s 212ms/step - loss: 0.1921 - accuracy: 0.9444 - val_loss: 3.7719 - val_accuracy: 0.7241
Epoch 118/150
18/18 [==============================] - 4s 215ms/step - loss: 0.2613 - accuracy: 0.9000 - val_loss: 2.2837 - val_ac

In [ ]:
score_adam = model_adam.evaluate(validation_set, verbose=0)
accuracy_adam = 100*score_adam[1]
print('Accuracy CNN Model with Adam Optimizer:', accuracy_adam)

In [ ]:
history_adam.history.keys()

## Model Loss:

In [ ]:
plt.figure(figsize=(12,7))
plt.xlabel("Epoch Number", fontsize=15)
plt.ylabel("Loss", fontsize=15)
plt.plot(history_adam.history["loss"], label= "training set")
plt.plot(history_adam.history["val_loss"], label= "validation set")
plt.legend()
plt.title("Model Loss", fontsize=20)

## Model Accuracy:

In [ ]:
plt.figure(figsize=(12, 7))
plt.xlabel("Epoch Number", fontsize=15)
plt.ylabel("Accuracy", fontsize= 15)
plt.plot(history_adam.history["accuracy"], label = "training set")
plt.plot(history_adam.history["val_accuracy"], label = "validation set")
plt.legend()
plt.title("Model Accuracy", fontsize=20)

## 2. Using RMSProp Optimizer:

In [ ]:
# Initialize the CNN model:
model_rms = Sequential()

# First Convolution Layer:
model_rms.add(Conv2D(filters=16, kernel_size=2, activation="relu", input_shape=[256, 256, 3]))

# First Pooling Layer:
model_rms.add(MaxPooling2D(pool_size=2))

# Second Convolution layer:
model_rms.add(Conv2D(filters= 32, kernel_size=2, activation="relu"))

# Second Pooling layer:
model_rms.add(MaxPooling2D(pool_size=2))

# Third Convolution Layer:
model_rms.add(Conv2D(filters= 32, kernel_size=2, activation="relu"))

# Third Pooling layer:
model_rms.add(MaxPooling2D(pool_size=2))

# Fourth Convolution Layer:
model_rms.add(Conv2D(filters=64, kernel_size=2, activation="relu"))

# Fourth Pooling Layer:
model_rms.add(MaxPooling2D(pool_size=2))

# Flattening:
model_rms.add(Flatten())

# Fully Connected Layer:
model_rms.add(Dense(128, activation="relu"))
# Deactivating 50% of neuron
model_rms.add(Dropout(0.5))

# Output Layer:
model_rms.add(Dense(3, activation="softmax"))

In [ ]:
# Layers:
model_rms.layers

In [ ]:
model_rms.summary()

## Compiling the Model:

In [ ]:
from keras.optimizers import RMSprop
model_rms.compile(loss= "categorical_crossentropy", optimizer = "RMSprop", metrics="accuracy")

In [ ]:
early_stopping =tf.keras.callbacks.EarlyStopping(monitor="val_loss",
                                                min_delta=0.0001,
                                                patience=20,
                                                verbose=1,
                                                mode="auto",
                                                baseline=None,
                                                restore_best_weights=False,
                                                start_from_epoch=0)

In [ ]:
early_stopping

In [ ]:
# Training the Model:
history_rms = model_rms.fit_generator(training_set,
                                      steps_per_epoch=len(training_set),
                                      epochs=150,
                                      validation_data=validation_set,
                                      validation_steps=len(validation_set))

Epoch 115/150
18/18 [==============================] - 4s 216ms/step - loss: 0.1966 - accuracy: 0.9111 - val_loss: 1.8251 - val_accuracy: 0.6207
Epoch 116/150
18/18 [==============================] - 4s 211ms/step - loss: 0.1906 - accuracy: 0.9111 - val_loss: 3.4237 - val_accuracy: 0.7586
Epoch 117/150
18/18 [==============================] - 4s 214ms/step - loss: 0.4506 - accuracy: 0.8667 - val_loss: 2.3255 - val_accuracy: 0.6552
Epoch 118/150
18/18 [==============================] - 4s 214ms/step - loss: 0.1473 - accuracy: 0.9444 - val_loss: 1.9090 - val_accuracy: 0.5172
Epoch 119/150
18/18 [==============================] - 4s 213ms/step - loss: 0.2178 - accuracy: 0.9333 - val_loss: 3.0895 - val_accuracy: 0.7931
Epoch 120/150
18/18 [==============================] - 4s 206ms/step - loss: 0.5437 - accuracy: 0.8556 - val_loss: 2.6139 - val_accuracy: 0.7586
Epoch 121/150
18/18 [==============================] - 4s 213ms/step - loss: 0.1752 - accuracy: 0.9333 - val_loss: 4.1641 - val_ac

In [ ]:
score_rms = model_rms.evaluate(validation_set, verbose=0)
accuracy_rms = 100*score_rms[1]
print("Accuracy CNN Model with RMSprop Optimizer:", accuracy_rms)

In [ ]:
history_rms.history.keys()

## Model Loss:

In [ ]:
plt.figure(figsize=size=(12,7))xlabelbelxlabel("Epoch Number", fontsizeylabel)
plt.ylabel("Loss", fontsize=1history_rms(history_rms.history["loss"], label = "training set")
plt.plot(history_rms.history["val_loss"], label = "validation set")
plt.legend()
plt.title("Model Loss", fontsize=20)

## Model Accuracy:

In [ ]:
plt.figure(figsize=(12,7))
plt.xlabel("Epoch Number", fontsize=15)
plt.ylabel("Accuracy", fontsize=15)
plt.plot(history_rms.history["accuracy"], label ="training set")
plt.plot(history_rms.history["val_accuracy"], label = "validation set")
plt.legend()
plt.title("Model Accuracy", fontsize=20)

# 3. Using SGD as Optimizer:


In [ ]:
# Initialize the CNN model:
model_sgd = Sequential()

# First Convolution Layer:
model_sgd.add(Conv2D(filters=16, kernel_size=2, activation="relu", input_shape=[256, 256, 3]))

# First Pooling Layer:
model_sgd.add(MaxPooling2D(pool_size=2))

# Second Convolution layer:
model_sgd.add(Conv2D(filters= 32, kernel_size=2, activation="relu"))

# Second Pooling layer:
model_sgd.add(MaxPooling2D(pool_size=2))

# Third Convolution Layer:
model_sgd.add(Conv2D(filters= 32, kernel_size=2, activation="relu"))

# Third Pooling layer:
model_sgd.add(MaxPooling2D(pool_size=2))

# Fourth Convolution Layer:
model_sgd.add(Conv2D(filters=64, kernel_size=2, activation="relu"))

# Fourth Pooling Layer:
model_sgd.add(MaxPooling2D(pool_size=2))

# Flattening:
model_sgd.add(Flatten())

# Fully Connected Layer:
model_sgd.add(Dense(128, activation="relu"))
# Deactivating 50% of neuron
model_sgd.add(Dropout(0.5))

# Output Layer:
model_sgd.add(Dense(3, activation="softmax"))

In [ ]:
# Layers:
model_sgd.layers

In [ ]:
model_sgd.summary()

 ## compiling the Model with SGD Optimizer:

In [ ]:
from keras.optimizers import SGD
model_sgd.compile(loss= "categorical_crossentropy", optimizer = "SGD", metrics="accuracy")

In [ ]:
# Training the Model:
history_sgd = model_sgd.fit_generator(training_set,
                                      steps_per_epoch=len(training_set),
                                      epochs=150,
                                      validation_data=validation_set,
                                      validation_steps=len(validation_set))

Epoch 115/150
18/18 [==============================] - 4s 195ms/step - loss: 0.6641 - accuracy: 0.6667 - val_loss: 0.5758 - val_accuracy: 0.8621
Epoch 116/150
18/18 [==============================] - 3s 194ms/step - loss: 0.6010 - accuracy: 0.7222 - val_loss: 0.5592 - val_accuracy: 0.8621
Epoch 117/150
18/18 [==============================] - 3s 196ms/step - loss: 0.6077 - accuracy: 0.7222 - val_loss: 0.6241 - val_accuracy: 0.8276
Epoch 118/150
18/18 [==============================] - 3s 192ms/step - loss: 0.8171 - accuracy: 0.6667 - val_loss: 1.9834 - val_accuracy: 0.3793
Epoch 119/150
18/18 [==============================] - 3s 189ms/step - loss: 0.7219 - accuracy: 0.6778 - val_loss: 0.6230 - val_accuracy: 0.6897
Epoch 120/150
18/18 [==============================] - 3s 190ms/step - loss: 0.6898 - accuracy: 0.7111 - val_loss: 0.5335 - val_accuracy: 0.8966
Epoch 121/150
18/18 [==============================] - 3s 194ms/step - loss: 0.7144 - accuracy: 0.6556 - val_loss: 0.3932 - val_ac

In [ ]:
score_sgd = model_sgd.evaluate(validation_set, verbose=0)
accuracy_sgd = 100*score_sgd[1]
print("Accuracy CNN Model with SGD Optimizer:", accuracy_sgd)

In [ ]:
history_sgd.history.keys()

## Model Loss:

In [ ]:
plt.figure(figsize=(12,7))
plt.xlabel('Epoch Number',fontsize=15)
plt.ylabel('Loss',fontsize=15)
plt.plot(history_sgd.history['loss'], label = 'training set')
plt.plot(history_sgd.history['val_loss'], label = 'validation set')
plt.legend()
plt.title("Model Loss",fontsize=20)

## Model Accuracy:

In [ ]:
plt.figure(figsize=(12,7))
plt.xlabel("Epoch Number", fontsize=15)
plt.ylabel("Accuracy", fontsize=15)
plt.plot(history_sgd.history["accuracy"], label ="training set")
plt.plot(history_sgd.history["val_accuracy"], label = "validation set")
plt.legend()
plt.title("Model Accuracy", fontsize=20)

## 4.Using Adagrad as Optimizer:

In [ ]:
# Initialize the CNN model:
model_adagrad = Sequential()

# First Convolution Layer:
model_adagrad.add(Conv2D(filters=16, kernel_size=2, activation="relu", input_shape=[256, 256, 3]))

# First Pooling Layer:
model_adagrad.add(MaxPooling2D(pool_size=2))

# Second Convolution layer:
model_adagrad.add(Conv2D(filters= 32, kernel_size=2, activation="relu"))

# Second Pooling layer:
model_adagrad.add(MaxPooling2D(pool_size=2))

# Third Convolution Layer:
model_adagrad.add(Conv2D(filters= 32, kernel_size=2, activation="relu"))

# Third Pooling layer:
model_adagrad.add(MaxPooling2D(pool_size=2))

# Fourth Convolution Layer:
model_adagrad.add(Conv2D(filters=64, kernel_size=2, activation="relu"))

# Fourth Pooling Layer:
model_adagrad.add(MaxPooling2D(pool_size=2))

# Flattening:
model_adagrad.add(Flatten())

# Fully Connected Layer:
model_adagrad.add(Dense(128, activation="relu"))
# Deactivating 50% of neuron
model_adagrad.add(Dropout(0.5))

# Output Layer:
model_adagrad.add(Dense(3, activation="softmax"))

In [ ]:
# Layers:
model_adagrad.layers

In [ ]:
model_adagrad.summary()

 ## compiling the Model with Adagrad Optimizer:

In [ ]:
from keras.optimizers import Adagrad
model_adagrad.compile(loss= "categorical_crossentropy", optimizer = "Adagrad", metrics="accuracy")

In [ ]:
# Training the Model:
history_adagrad = model_adagrad.fit_generator(training_set,
                                      steps_per_epoch=len(training_set),
                                      epochs=150,
                                      validation_data=validation_set,
                                      validation_steps=len(validation_set))

Epoch 115/150
18/18 [==============================] - 3s 187ms/step - loss: 0.9272 - accuracy: 0.5444 - val_loss: 0.8883 - val_accuracy: 0.6552
Epoch 116/150
18/18 [==============================] - 3s 194ms/step - loss: 0.9710 - accuracy: 0.5222 - val_loss: 0.9042 - val_accuracy: 0.5517
Epoch 117/150
18/18 [==============================] - 3s 190ms/step - loss: 0.9393 - accuracy: 0.5667 - val_loss: 0.8932 - val_accuracy: 0.5862
Epoch 118/150
18/18 [==============================] - 3s 187ms/step - loss: 0.9752 - accuracy: 0.5000 - val_loss: 0.9108 - val_accuracy: 0.5172
Epoch 119/150
18/18 [==============================] - 3s 191ms/step - loss: 0.9493 - accuracy: 0.5222 - val_loss: 0.9361 - val_accuracy: 0.5517
Epoch 120/150
18/18 [==============================] - 3s 186ms/step - loss: 0.9610 - accuracy: 0.5333 - val_loss: 0.9005 - val_accuracy: 0.5517
Epoch 121/150
18/18 [==============================] - 3s 187ms/step - loss: 0.9633 - accuracy: 0.5444 - val_loss: 0.8780 - val_ac

In [ ]:
score_adagrad = model_adagrad.evaluate(validation_set, verbose=0)
accuracy_adagrad = 100*score_adagrad[1]
print("Accuracy CNN Model with Adagrad Optimizer:", accuracy_adagrad)

In [ ]:
history_adagrad.history.keys()

## Model Loss:

In [ ]:
plt.figure(figsize=(12,7))
plt.xlabel('Epoch Number',fontsize=15)
plt.ylabel('Loss',fontsize=15)
plt.plot(history_adagrad.history['loss'], label = 'training set')
plt.plot(history_adagrad.history['val_loss'], label = 'validation set')
plt.legend()
plt.title("Model Loss",fontsize=20)

## Model Accuracy:

In [ ]:
plt.figure(figsize=(12,7))
plt.xlabel("Epoch Number", fontsize=15)
plt.ylabel("Accuracy", fontsize=15)
plt.plot(history_adagrad.history["accuracy"], label ="training set")
plt.plot(history_adagrad.history["val_accuracy"], label = "validation set")
plt.legend()
plt.title("Model Accuracy", fontsize=20)

# RESULT:

## Comparing the optimizers for the best accuracy:

In [ ]:
accuracy = [accuracy_adam,accuracy_rms,accuracy_sgd,accuracy_adagrad]
optimizers = ["ADAM","RMSprop","SGD","ADAGRAD"]    

for i in range(len(optimizers)):
    print("The acuuracy achieved using "+optimizers[i]+" is: "+str(accuracy[i])+" %")

In [ ]:
plt.figure(figsize=(10,6))
plt.xlabel("Optimizers")
plt.ylabel("Accuracy")
ax = sns.barplot(x=optimizers, y=accuracy)
for label in ax.containers:
    ax.bar_label(label)
plt.tight_layout()
plt.tick_params(labelsize=14)

# Conclusion:
To conclude, the application of Convolutional Neural Networks (CNNs) in predicting rice leaf diseases has yielded promising outcomes, underscoring the pivotal role of optimizer selection in achieving optimal accuracy. The analysis has led to several key insights:

**Optimizers and Accuracy:** Among the four optimizers scrutinized, ADAM emerged as the top performer, attaining an impressive accuracy rate of 72.41%. This underscores the efficacy of ADAM in fine-tuning the CNN model's parameters, resulting in superior disease prediction for rice leaves.

**Alternative Optimizer:** Stochastic Gradient Descent (SGD), the second-best optimizer, demonstrated robust performance with an accuracy of 86.20%. This underscores the adaptability of CNNs, showcasing their capability to deliver reliable results with different optimization algorithms.

**RMSProp Performance:** Although not reaching the levels of ADAM and SGD, RMSProp exhibited reasonable accuracy at 82.75%. This suggests its viability as a choice, particularly when computational resources or training time are constrained.

**Performance Variation:** ADAGRAD lagged significantly behind, yielding an accuracy of 62.06%. This signals that, for this specific task, ADAGRAD may not be the most suitable option, prompting a reconsideration of its use in favor of more effective optimization algorithms.

**Optimal Choice:** Considering all optimizers, SGD emerges as the preferred optimizer for disease prediction in rice leaves. Its consistently high accuracy across multiple experiments establishes its reliability and appropriateness for this specific agricultural task.

In summary, the synergy of CNNs with a well-chosen optimizer, such as SGD, proves instrumental in accurately predicting rice leaf diseases. The attained high accuracy underscores the potential of deep learning techniques in agriculture, offering a means to identify and address crop diseases promptly, ultimately contributing to enhanced crop yield and global food security. Ongoing research, coupled with hyperparameter optimization, holds promise for further refining CNN models and bolstering their efficacy in agricultural applications.

# Challenges Encountered
**Data Loading Complexity:** One of the initial challenges encountered was the intricacy of loading data, particularly when dealing with extensive datasets or diverse file formats. Overcoming issues related to correct file paths, handling missing or corrupted data, and ensuring data consistency required the implementation of robust data loading mechanisms and effective error-handling strategies.

**Model Training Time Constraints:** The computational intensity and time-consuming nature of training deep learning models, notably CNNs, posed a significant challenge. Large datasets and complex architectures contributed to prolonged training times. Mitigating this challenge involved leveraging efficient hardware, such as GPUs or TPUs, and exploring distributed training methods to expedite the model training process.

**Optimization Algorithm Variability:** Notable variability in model performance, specifically in accuracy, was observed when experimenting with different optimizers, hyperparameters, and random weight initialization. Addressing this challenge involved conducting multiple runs with each optimizer and reporting average performance to provide a more stable basis for comparison.

**Optimal Optimizer Selection:** Selecting the most suitable optimizer proved to be a nuanced task, given the diverse outcomes produced by different optimization algorithms. To address this challenge systematically, thorough hyperparameter tuning, including grid search or random search, was employed for each optimizer. Additionally, cross-validation was instrumental in assessing model stability and generalization performance across various optimizers.

**Resource Limitations:** The inherent resource intensity of training deep learning models, coupled with the need for powerful hardware and substantial memory, presented a notable constraint. Overcoming this challenge required strategic resource allocation and optimization, ensuring that the experiments conducted were within the available computational capabilities.

In summary, the journey involved in building and training deep learning models for rice leaf disease prediction presented multifaceted challenges. Overcoming complexities related to data loading, training time, optimizer variability, and resource constraints demanded a combination of technical acumen, meticulous coding practices, access to suitable hardware, and methodical experimentation. Rigorous testing and evaluation methodologies played a pivotal role in making well-informed decisions about model architectures, hyperparameters, and optimizers to achieve optimal performance for the given agricultural task.






